### PR TITLE
Pass event_builder in as a dependency

### PR DESCRIPTION
**Pull request by robjacoby**
If an event that is being processed does not conform to the specific
structure that the EventBuilder is expecting, the event store will fail.

By passing in the EventBuilder as a dependency, we are able to build an
Event Framework Event from any stored event structure.

Interestingly, ruby allows you to reference a given kwarg as a parameter
in a different default kwarg value, as in
```
event_builder: EventBuilder.new(event_type_resolver: event_type_resolver)
```

Culture Amp Internal Context:

I want to be able to use this gem to process events from Influx. This change will allow me to build an `EventFramework::Event` from the structure in said events.

### CODE REVIEW

**Review by twe4ked — DISMISSED** [review]
LGTM. I just had a quick look at the `EventBuilder` class though and there's quite a bit going on there. Things like upcasting event bodies, upcasting metdata, and instantiating metadata. That's quite a bit of stuff to ensure any external users are doing. Maybe that's okay? Another option could be to introduce some kind of pipeline so you can override specific parts.

Thoughts?

Approving anyway to unblock you, but please have a think.

**Comment by robjacoby** [comment]
Yeah for sure. Right now there's a LOT of assumptions this gem makes, from the naming conventions in classes, to required keys in metadata, and event structure.

I think as we go, ensuring there are options to inject different dependencies will be key. Alternatively, we may decide that it's ok to not support a plethora of different structures.

For instance with the event builder, maybe we don't _need_ to worry about upcasting, so perhaps it's ok that our version doesn't do that?

**Review by timriley — APPROVED** [review]
I think this is good as a way to unblock you and the team, @robjacoby 👍 

I do think @twe4ked is right that we should also use this as a prompt to review configurability in general. For example, one thing that jumps out to me even here is that it might be nice to separate the `event_builder` from the `event_type_resolver` – I could well imagine an application wanting to provide one but not the other, but in this current arrangement, as soon as you want your own event_builder, you're forced to worry about _also_ wiring up the event_type_resolver.

Anyway, this is a good interim step at least!

**Review by timriley — APPROVED** [review]

**Comment by twe4ked** [comment]
Mind rebasing out the merge commit before merge to keep the history tidy?

**Comment by robjacoby** [comment]
Done @twe4ked :)

**Review by twe4ked — APPROVED** [review]
👍 

Remember to run the specs locally as there is no CI set up still.

**Pull request merged**
